### PR TITLE
iot: add mqtt gateway snippets + test

### PR DESCRIPTION
**Pull request by hongalex**
Added samples for Gateway features including
- Having a gateway send data on behalf of a bound device
- Having a gateway listen for messages on a bound device's topic (state, config, errors)

### CODE REVIEW

**Comment by hongalex** [comment]
I have excluded tests for `create_jwt` mostly because it doesn't test functionality for IoT Core and because we don't have it for other samples, but open to discuss this.

File: iot/mqttsnippets/attach_detach_device.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2018 Google LLC. All rights reserved.
```
Comment (gguuss) [inline iot/mqttsnippets/attach_detach_device.go]: Nit, if it's a new file, 2019

File: iot/mqttsnippets/attach_detach_device.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	"fmt"
+
+	"github.com/eclipse/paho.mqtt.golang"
+)
+
+// [START iot_attach_device]
+
+// attaches a devices to a gateway.
```
Comment (gguuss) [inline iot/mqttsnippets/attach_detach_device.go]: attaches a device to a gateway

File: iot/mqttsnippets/attach_detach_device.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+		return token.Error()
+	}
+	return nil
+}
+
+// [END iot_attach_device]
+
+// [START iot_detach_device]
+
+// detaches a devices to a gateway.
```
Comment (gguuss) [inline iot/mqttsnippets/attach_detach_device.go]: detaches a device from a gateway

Comment (gguuss) [reply iot/mqttsnippets/attach_detach_device.go]: Does this (and attach) return nil on success?

File: iot/mqttsnippets/gateway_send_data_test.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+	}
+	client, err := cloudiot.New(httpClient)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// [END iot_get_client]
```
Comment (gguuss) [inline iot/mqttsnippets/gateway_send_data_test.go]: Any reason for the include code blocks in the test?

File: iot/mqttsnippets/attach_detach_device.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		return token.Error()
+	}
+	return nil
+}
+
+// [END iot_attach_device]
+
+// [START iot_detach_device]
+
+// detaches a device from a gateway.
```
Comment (tbpg) [inline iot/mqttsnippets/attach_detach_device.go]:
```suggestion
// detachDevice detaches a device from a gateway.
```

File: iot/mqttsnippets/create_jwt.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"errors"
+	"io/ioutil"
+	"time"
+
+	jwt "github.com/dgrijalva/jwt-go"
+)
+
+// createJWT creates a Cloud IoT Core JWT for the given project id.
+// algorithm can be one of ["RSA256", "ES256"]
```
Comment (tbpg) [inline iot/mqttsnippets/create_jwt.go]: Period at the end.

File: iot/mqttsnippets/create_jwt.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+func createJWT(projectID string, privateKeyPath string, algorithm string, expiresMin int) (string, error) {
```
Comment (tbpg) [inline iot/mqttsnippets/create_jwt.go]: Make expiresMin a `time.Duration` and use `time.Now().Add(expiration).Unix()`? Alternatively, make the expiration the `time.Time` it should expire?

File: iot/mqttsnippets/create_jwt.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+	switch algorithm {
+	case "RS256":
+		privKey, _ := jwt.ParseRSAPrivateKeyFromPEM(keyBytes)
+		return token.SignedString(privKey)
+	case "ES256":
+		privKey, _ := jwt.ParseECPrivateKeyFromPEM(keyBytes)
+		return token.SignedString(privKey)
+	default:
+		return "", errors.New("Cannot find JWT algorithm. Specify 'ES256' or 'RS256'")
```
Comment (tbpg) [inline iot/mqttsnippets/create_jwt.go]: This works, but I might move the final return out of the `switch` and drop the `default` branch to make it more clear that's what happens if we haven't returned yet.

File: iot/mqttsnippets/gateway_send_data.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	"io"
+	"math/rand"
+	"time"
+
+	"github.com/eclipse/paho.mqtt.golang"
+)
+
+// [START iot_send_data_from_bound_device]
+
+// Sends data on behalf of a bound device
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data.go]:
```suggestion
// sendDataFromBoundDevice sends data on behalf of a bound device.
```

File: iot/mqttsnippets/gateway_send_data.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+		fmt.Fprintln(w, "Failed to connect client")
+		return
+	}
+
+	if err := attachDevice(deviceID, client, ""); err != nil {
+		fmt.Fprintf(w, "Failed to attach device %s\n", err)
+		return
+	}
+
+	time.Sleep(5 * time.Second)
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data.go]: Add a comment explaining the sleep?

File: iot/mqttsnippets/gateway_send_data.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+	// Create and connect a client using the above options.
+	client := mqtt.NewClient(opts)
+	if token := client.Connect(); token.Wait() && token.Error() != nil {
+		fmt.Fprintln(w, "Failed to connect client")
+		return
+	}
+
+	if err := attachDevice(deviceID, client, ""); err != nil {
+		fmt.Fprintf(w, "Failed to attach device %s\n", err)
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data.go]: Return an error?

File: iot/mqttsnippets/gateway_send_data.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	time.Sleep(5 * time.Second)
+
+	gatewayStateTopic := fmt.Sprintf("/devices/%s/state", gatewayID)
+	deviceStateTopic := fmt.Sprintf("/devices/%s/state", deviceID)
+
+	gatewayInitPayload := fmt.Sprintf("Starting gateway at time: %d", time.Now().Unix())
+	if token := client.Publish(gatewayStateTopic, 1, false, gatewayInitPayload); token.Wait() && token.Error() != nil {
+		fmt.Fprintln(w, "Failed to publish initial gateway payload")
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data.go]: Return the error?

File: iot/mqttsnippets/gateway_send_data.go [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+		return
+	}
+
+	for i := 1; i <= numMessages; i++ {
+		if shouldBackoff {
+			if backoffTime > maximumBackoffTime {
+				fmt.Fprintln(w, "Exceeded max backoff time.")
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data.go]: Return error.

File: iot/mqttsnippets/gateway_send_data.go [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+				return
+			}
+
+			waitTime := backoffTime + rand.Intn(1000)/1000.0
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data.go]: I'm not familiar with backoff strategies - why is this added?

File: iot/mqttsnippets/gateway_send_data.go [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+	deviceStateTopic := fmt.Sprintf("/devices/%s/state", deviceID)
+
+	gatewayInitPayload := fmt.Sprintf("Starting gateway at time: %d", time.Now().Unix())
+	if token := client.Publish(gatewayStateTopic, 1, false, gatewayInitPayload); token.Wait() && token.Error() != nil {
+		fmt.Fprintln(w, "Failed to publish initial gateway payload")
+		return
+	}
+
+	for i := 1; i <= numMessages; i++ {
+		if shouldBackoff {
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data.go]: `shouldBackoff` feels weird to me. Would it be enough to check if `backoffTime > 0`?

File: iot/mqttsnippets/gateway_send_data.go [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+			backoffTime *= 2
+			client = mqtt.NewClient(opts)
+		}
+
+		deviceStatePayload := fmt.Sprintf("%s, #%d", payload, i)
+
+		fmt.Fprintf(w, "Publishing message: %s to %s\n", deviceStatePayload, deviceStateTopic)
+
+		if token := client.Publish(deviceStateTopic, 1, false, payload); token.Wait() && token.Error() != nil {
+			fmt.Fprintln(w, "Failed to publish payload to device state topic")
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data.go]: Return error.

File: iot/mqttsnippets/gateway_send_data_test.go [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+	"golang.org/x/oauth2/google"
+	cloudiot "google.golang.org/api/cloudiot/v1"
+)
+
+var privateKeyRSA = os.Getenv("GOLANG_SAMPLES_IOT_PRIV")
+var pubKeyRSA = os.Getenv("GOLANG_SAMPLES_IOT_PUB")
+var region = "us-central1"
+
+// returns a v1 UUID for a resource: e.g. topic, registry, gateway, device
+func createIDForTest(resource string) string {
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data_test.go]: This is in a test file. So, I don't think we need the `ForTest`.

File: iot/mqttsnippets/gateway_send_data_test.go [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+	"github.com/google/uuid"
+	"golang.org/x/oauth2/google"
+	cloudiot "google.golang.org/api/cloudiot/v1"
+)
+
+var privateKeyRSA = os.Getenv("GOLANG_SAMPLES_IOT_PRIV")
+var pubKeyRSA = os.Getenv("GOLANG_SAMPLES_IOT_PUB")
+var region = "us-central1"
+
+// returns a v1 UUID for a resource: e.g. topic, registry, gateway, device
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data_test.go]:
```suggestion
// createID returns a v1 UUID for a resource: e.g. topic, registry, gateway, device.
```

File: iot/mqttsnippets/gateway_send_data_test.go [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+func createIDForTest(resource string) string {
+	uuid, err := uuid.NewRandom()
+	if err != nil {
+		log.Fatalf("Could not generate uuid: %v", err)
+	}
+	id := fmt.Sprintf("golang-test-%s-%s", resource, uuid.String())
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data_test.go]:
```suggestion
	return fmt.Sprintf("golang-test-%s-%s", resource, uuid.String())
```

File: iot/mqttsnippets/gateway_send_data_test.go [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+	client, err := cloudiot.New(httpClient)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// createRegistry creates a IoT Core device registry associated with a PubSub topic
+func createRegistry(w io.Writer, projectID string, region string, registryID string, topicName string) (*cloudiot.DeviceRegistry, error) {
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data_test.go]: Are there samples for all of these helpers? Or they're in another package?

File: iot/mqttsnippets/gateway_send_data_test.go [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+func TestSendDataFromBoundDevice(t *testing.T) {
+	projectID := testutil.SystemTest(t).ProjectID
+
+	registryID := "golang-iot-test-registry"
+	gatewayID := createIDForTest("gateway")
+	deviceID := createIDForTest("device")
+
+	testutil.Retry(t, 1, 10*time.Second, func(r *testutil.R) {
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data_test.go]: Might be better to have a retry for each part that could fail, rather than the end-to-end test. As is, if the last step fails, you have to redo all of the others.

File: iot/mqttsnippets/gateway_send_data_test.go [hunk 22]
```diff
@@ -0,0 +1,284 @@
...
+		want := fmt.Sprintf("Publishing message: %s", "test")
+		if got := buf.String(); !strings.Contains(got, want) {
+			r.Errorf("SendDataFromBoundDevice got %s, want substring %q", got, want)
+		}
+	})
+
+	// cleanup
+	unbindDeviceFromGateway(ioutil.Discard, projectID, region, registryID, gatewayID, deviceID)
+	deleteDevice(ioutil.Discard, projectID, region, registryID, deviceID)
+	deleteDevice(ioutil.Discard, projectID, region, registryID, gatewayID)
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data_test.go]: deleteDevice on a gateway? Confirming this is expected.

File: iot/mqttsnippets/gateway_send_data_test.go [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+		sendDataFromBoundDevice(buf, projectID, region, registryID, gatewayID, deviceID, privateKeyRSA, "RS256", 2, "test")
+
+		want := fmt.Sprintf("Publishing message: %s", "test")
+		if got := buf.String(); !strings.Contains(got, want) {
+			r.Errorf("SendDataFromBoundDevice got %s, want substring %q", got, want)
+		}
+	})
+
+	// cleanup
+	unbindDeviceFromGateway(ioutil.Discard, projectID, region, registryID, gatewayID, deviceID)
```
Comment (tbpg) [inline iot/mqttsnippets/gateway_send_data_test.go]: Might be good to have a retry for these, too.

File: iot/mqttsnippets/subscribe_gateway_device_topic.go [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/eclipse/paho.mqtt.golang"
+)
+
+// [START iot_subuscribe_config_bound_device]
```
Comment (tbpg) [inline iot/mqttsnippets/subscribe_gateway_device_topic.go]: If it's in a single file, the region should include the imports.

File: iot/mqttsnippets/subscribe_gateway_device_topic.go [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+
+// subscribeGatewayToDeviceTopic creates a gateway client that subscribes to a topic of a bound device.
+// Currently supported topics include: "config", "state", "commands", "errors"
+func subscribeGatewayToDeviceTopic(w io.Writer, projectID string, region string, registryID string, gatewayID string, deviceID string, privateKeyPath string, algorithm string, clientDuration int, topic string) {
```
Comment (tbpg) [inline iot/mqttsnippets/subscribe_gateway_device_topic.go]: Should return an error.

File: iot/mqttsnippets/subscribe_gateway_device_topic_test.go [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+func TestSubscribeGatewayToDeviceTopic(t *testing.T) {
+	projectID := testutil.SystemTest(t).ProjectID
+
+	registryID := "golang-iot-test-registry"
+	gatewayID := createIDForTest("gateway")
+	deviceID := createIDForTest("device")
+
+	testutil.Retry(t, 1, 10*time.Second, func(r *testutil.R) {
```
Comment (tbpg) [inline iot/mqttsnippets/subscribe_gateway_device_topic_test.go]: Same retry comment.

File: iot/mqttsnippets/attach_detach_device.go (hunk 3 above)
Comment (tbpg) [reply iot/mqttsnippets/attach_detach_device.go]: Yes, returning `nil` when there is no error is idiomatic and indicates success.

File: iot/mqttsnippets/gateway_send_data_test.go (hunk 23 above)
Comment (hongalex) [reply iot/mqttsnippets/gateway_send_data_test.go]: Done.

File: iot/mqttsnippets/subscribe_gateway_device_topic_test.go [hunk 27]
```diff
@@ -44,10 +44,10 @@ func TestSubscribeGatewayToDeviceTopic(t *testing.T) {
 	projectID := testutil.SystemTest(t).ProjectID
 
 	registryID := "golang-iot-test-registry"
-	gatewayID := createIDForTest("gateway")
-	deviceID := createIDForTest("device")
+	gatewayID := createID("gateway")
+	deviceID := createID("device")
 
-	testutil.Retry(t, 1, 10*time.Second, func(r *testutil.R) {
+	testutil.Retry(t, 5, 5*time.Second, func(r *testutil.R) {
```
Comment (tbpg) [inline iot/mqttsnippets/subscribe_gateway_device_topic_test.go]: Sorry, I think I was unclear. Should all of these `create` calls be in separate `testutil.Retry` blocks? Try to create the gateway, then try to create device, then try to bind them. And so on. That way, if everything succeeds except binding the device, the test doesn't have to create the gateway/device again. Or maybe `createGateway` and `createDevice` do nothing if they already exist?

Comment (hongalex) [reply iot/mqttsnippets/subscribe_gateway_device_topic_test.go]: So as it stands right now, you're right, trying to create a device that already exists will throw an error. I could rewrite the existing code to account for this, but I think wrapping each call in separate `testutil.Retry` probably makes more sense.

File: iot/mqttsnippets/gateway_send_data_test.go (hunk 20 above)
Comment (tbpg) [reply iot/mqttsnippets/gateway_send_data_test.go]: Could we put them in the same directory/package? I think the cost of having samples in the same directory is lower than copying the helpers. Feel free to disagree. It's OK the manager is `package main` - these could be, too.

Comment (hongalex) [reply iot/mqttsnippets/gateway_send_data_test.go]: Potentially worried about scaling long term. If we continue to add samples at the rate of 1 file per sample, this directory could get very big very fast. Willing to entertain the idea though since it decreases duplication, but I think @broady @gguuss might have more ideas.

Comment (gguuss) [reply iot/mqttsnippets/gateway_send_data_test.go]: Anecdotally, we have put one snippet per file in PHP; there are a good number of files already but it's not unmaintainable. If we wanted to coalesce the snippets to each feature, that's one way we could group them and reduce the file count.  Longer term, I want to ship e2e demos showing the features, which could be put in the main package.

Comment (hongalex) [reply iot/mqttsnippets/gateway_send_data_test.go]: I'd be willing to move to a one directory snippet, but as of right now, the refactor into one directory would cause a lot of issues. Basically it would involve updating the docs to point to a different directory (since I can't just move MQTT samples into the `manager` directory as well as fixing the tests as well. Also, it would make this specific PR very large since there will be a good amount of renaming files and such. I would prefer if we pushed this to a later time and keep it the way it is for now.

Comment (hongalex) [reply iot/mqttsnippets/gateway_send_data_test.go]: Also, having a single directory for device management and MQTT samples would be hard for pages like this: https://cloud.google.com/iot/docs/samples/mqtt-samples which have links that point to a specific directory.

Comment (tbpg) [reply iot/mqttsnippets/gateway_send_data_test.go]: Got it. Yeah, if the directory is linked from a doc, it's probably good to keep it clean.
